File: qqbot-client/src/main/java/com/mieai/qqbot/client/QqTextMessageRequest.java
package com.mieai.qqbot.client;

import java.util.Objects;
import java.util.Optional;

/** Validated text request used by the QQ OpenAPI sender. */
public record QqTextMessageRequest(
        QqMessageTargetType targetType,
        String targetId,
        String content,
        Optional<String> replyMessageId,
        Optional<String> replyEventId,
        int messageSequence) {
    public QqTextMessageRequest {
        Objects.requireNonNull(targetType, "targetType must not be null");
        requirePathToken(targetId, "targetId");
        Objects.requireNonNull(content, "content must not be null");
        if (content.isBlank() || content.codePoints().anyMatch(QqTextMessageRequest::unsupportedControl)) {
            throw new IllegalArgumentException("content must be non-blank and free of control characters");
        }
        if (content.codePointCount(0, content.length()) > 4000) throw new IllegalArgumentException("content is too long");
        Objects.requireNonNull(replyMessageId, "replyMessageId must not be null");
        Objects.requireNonNull(replyEventId, "replyEventId must not be null");
        if (messageSequence < 1) throw new IllegalArgumentException("messageSequence must be positive");
    }

    private static boolean unsupportedControl(int value) {
        return Character.isISOControl(value) && value != '\n' && value != '\r' && value != '\t';
    }

    private static void requirePathToken(String value, String name) {
        if (value == null || value.isBlank() || !value.equals(value.strip())
                || value.codePoints().anyMatch(Character::isWhitespace)
                || value.indexOf('/') >= 0 || value.indexOf('\\') >= 0
                || value.indexOf('?') >= 0 || value.indexOf('#') >= 0) {
            throw new IllegalArgumentException(name + " is invalid");
        }
    }
}

File: qqbot-client/src/main/kotlin/com/mieai/qqbot/client/QqTextMessageRequest.kt
package com.mieai.qqbot.client

/** Validated text request used by the QQ OpenAPI sender. */
data class QqTextMessageRequest(
    val targetType: QqMessageTargetType,
    val targetId: String,
    val content: String,
    val replyMessageId: String?,
    val replyEventId: String?,
    val messageSequence: Int,
) {
    init {
        requirePathToken(targetId, "targetId")
        require(!content.isBlank() && content.codePoints().noneMatch(::unsupportedControl)) {
            "content must be non-blank and free of control characters"
        }
        require(content.codePointCount(0, content.length) <= 4000) { "content is too long" }
        require(messageSequence >= 1) { "messageSequence must be positive" }
    }

    companion object {
        private fun unsupportedControl(value: Int): Boolean =
            Character.isISOControl(value) && value != '\n'.code && value != '\r'.code && value != '\t'.code

        private fun requirePathToken(value: String, name: String) {
            require(!value.isBlank() && value == value.trim() &&
                value.codePoints().noneMatch(Character::isWhitespace) &&
                '/' !in value && '\\' !in value && '?' !in value && '#' !in value) {
                "$name is invalid"
            }
        }
    }
}
